shared-libs/message-utils/src/index.js

```diff
@@ -96,77 +96,128 @@ const applyPhoneFilters = function(config, phone) {
+
+  return phone || (default_to_sender && from) || recipient;
+};
+
+const resolveRecipient = function(context, recipient) {
+  if (!recipient) {
+    return null;
+  }
+
+  const resolvers = [
+    {
+      match: r => r === 'reporting_unit',
+      resolve: () => context.from || context.contact?.phone,
+    },
+    {
+      match: r => r.startsWith('ancestor:'),
+      resolve: r => {
+        const type = r.split(':')[1];
+        return (
+          getParentPhone(context.patient, type) ||
+          getParentPhone(context.place, type) ||
+          getParentPhone(context, type)
+        );
+      },
+    },
+    {
+      match: r => r.startsWith('link:'),
+      resolve: r => {
+        const tag = r.split(':')[1];
+        return (
+          getLinkedPhone(context.patient, tag) ||
+          getLinkedPhone(context.place, tag) ||
+          getLinkedPhone(context.contact, tag) ||
+          getParentPhone(context.patient, tag) ||
+          getParentPhone(context.place, tag) ||
+          getParentPhone(context.contact, tag)
+        );
+      },
+    },
+    {
+      match: r => r === 'parent',
+      resolve: () => resolveAncestor(context, 2),
+    },
+    {
+      match: r => r === 'grandparent',
+      resolve: () => resolveAncestor(context, 3),
+    },
+    {
+      match: r => r === 'clinic',
+      resolve: () => getClinicPhone(context.patient) ||
+        getClinicPhone(context.place) ||
+        getClinicPhone(context) ||
+        context.contact?.phone,
+    },
+    {
+      match: r => r === 'health_center',
+      resolve: () => getHealthCenterPhone(context.patient) ||
+        getHealthCenterPhone(context.place) ||
+        getHealthCenterPhone(context),
+    },
+    {
+      match: r => r === 'district',
+      resolve: () => getDistrictPhone(context.patient) ||
+        getDistrictPhone(context.place) ||
+        getDistrictPhone(context),
+    },
+    {
+      match: r => context.fields?.[r],
+      resolve: r => context.fields[r],
+    },
+    {
+      match: r => context[r],
+      resolve: r => context[r],
+    },
+    {
+      match: r => r.includes('.'),
+      resolve: r => objectPath.get(context, r),
+    },
+    {
+      match: r => phoneNumber.validate({}, r),
+      resolve: r => r,
+    }
+  ];
+
+  for (const rule of resolvers) {
+    if (rule.match(recipient)) {
+      return rule.resolve(recipient);
+    }
+  }
+
+  return null;
+};
+
+const resolveAncestor = function(context, levels) {
+  let node = context.patient || context.place || context;
+  node = node.parent ? node : node.contact;
+
+  for (let i = 0; i < levels; i++) {
+    if (!node?.parent) {
+      return null;
+    }
+    node = node.parent;
   }
-  return phone || from || recipient;
+
+  return node.contact?.phone || null;
 };
 
 const getPhone = function(config, context, recipient) {
-  let phone = getRecipient(context, recipient);
+  let phone = getRecipient(context, recipient, config?.sms?.default_to_sender ?? true);
```

Comment:
I think I would rather pass the whole config to the `getRecipient` function and get the value of whether to default to sender within that function. it's a cleaner separation of concerns.

Reply:
I think it’s cleaner for the caller to resolve the `defaultToSender` boolean before invoking getRecipient.  It’s better for SoC: the caller decides the boolean, and getRecipient focuses on resolving a phone number.

If we start passing the whole config object, the function becomes tied to that structure. Keeping it decoupled feels like the safer design.